src/NearestNeighbors.jl

```diff
@@ -23,7 +23,6 @@ abstract type NNTree{V <: AbstractVector,P <: PreMetric} end
 
 const NonweightedMinowskiMetric = Union{Euclidean,Chebyshev,Cityblock,Minkowski}
 const WeightedMinowskiMetric = Union{WeightedEuclidean,WeightedCityblock,WeightedMinkowski}
```

Comment:
Are these used anywhere else than in the line below? If not, you may want to delete them.

Reply:
Yes, they are used:
```julia
# path: /src/evaluation.jl

@inline eval_pow(::MinkowskiMetric, s) = abs(s)
@inline eval_pow(::Euclidean, s) = abs2(s)
@inline eval_pow(::WeightedEuclidean, s) = abs2(s)
@inline eval_pow(d::Minkowski, s) = abs(s)^d.p

@inline eval_diff(::NonweightedMinowskiMetric, a, b, dim) = a - b
@inline eval_diff(::Chebyshev, ::Any, b, dim) = b
@inline eval_diff(m::WeightedMinowskiMetric, a, b, dim) = m.weights[dim] * (a-b)
```
I was thinking about how we can remove them and the ideas that came up were:
* `exponent(d::MinkowskiMetric)` return the `p` parameter of a Minkowski metric;
* `isweighted(d)` checks if metric `d` is weighted;
* `weights(d)` return the `weights` parameter of `d`;

Maybe this last function can be replaced by the `Distances.parameters` function, but I think this function is not public.

With these new APIs, these functions would be implemented as follows:
```julia
@inline eval_pow(m::MinkowskiMetric, s) = abs(s)^exponent(m)
# simple optimization
@inline eval_pow(::Euclidean, s) = abs2(s)
@inline eval_pow(::WeightedEuclidean, s) = abs2(s)

@inline eval_diff(m::MinkowskiMetric, a, b, dim) = isweighted(m) ? weights(m)[dim] * (a-b) : (a-b)
# simple optimization
@inline eval_diff(::Chebyshev, a, b, dim) = b
```

Reply:
What's the problem with defining these functions for your own metric?

Reply:
I see, so the idea is that new metrics define these functions. Makes sense.